### PR TITLE
bugfix: DataSourceProxyHolder#putDataSource deadlock

### PR DESCRIPTION
**Pull request by spilledyear**
#3567 
修复自动数据源代理 因 ConcurrentHashMap.computeIfAbsent 导致的死锁问题

### CODE REVIEW

File: spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java [hunk 1]
```diff
@@ -15,26 +15,26 @@
...
-import javax.sql.DataSource;
-import java.util.concurrent.ConcurrentHashMap;
-
-import io.seata.common.util.CollectionUtils;
 import io.seata.core.model.BranchType;
 import io.seata.rm.datasource.DataSourceProxy;
 import io.seata.rm.datasource.SeataDataSourceProxy;
 import io.seata.rm.datasource.xa.DataSourceProxyXA;
 
+import javax.sql.DataSource;
```
Comment (a364176773) [inline spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java:23]: Java package should be at the top level

File: spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java [hunk 2]
```diff
@@ -71,15 +71,28 @@ public SeataDataSourceProxy putDataSource(DataSource dataSource, BranchType data
...
         } else {
             originalDataSource = dataSource;
         }
-        return CollectionUtils.computeIfAbsent(this.dataSourceProxyMap, originalDataSource,
-                BranchType.XA == dataSourceProxyMode ? DataSourceProxyXA::new : DataSourceProxy::new);
+
+        SeataDataSourceProxy dsProxy = dataSourceProxyMap.get(originalDataSource);
+        if (dsProxy == null) {
+            synchronized (dataSourceProxyMap) {
+                if ((dsProxy = dataSourceProxyMap.get(originalDataSource)) == null) {
```
Comment (a364176773) [inline spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java:86]: dsProxy = dataSourceProxyMap.get(originalDataSource
if(dsProxy==null){
}

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java [hunk 3]
```diff
@@ -28,6 +25,9 @@
 import org.springframework.aop.IntroductionInfo;
 import org.springframework.beans.BeanUtils;
 
+import javax.sql.DataSource;
```
Comment (a364176773) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java:28]: Java package should be at the top level

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java [hunk 4]
```diff
@@ -53,15 +53,14 @@ public SeataAutoDataSourceProxyAdvice(String dataSourceProxyMode) {
...
-                && dataSourceProxyMode != RootContext.getBranchType()) {
+        if (!RootContext.requireGlobalLock() && dataSourceProxyMode != RootContext.getBranchType()) {
             return invocation.proceed();
         }
 
         Method method = invocation.getMethod();
         Object[] args = invocation.getArguments();
         Method m = BeanUtils.findDeclaredMethod(dataSourceProxyClazz, method.getName(), method.getParameterTypes());
-        if (m != null) {
+        if (m != null && !Object.class.equals(method.getDeclaringClass())) {
```
Comment (a364176773) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java:63]: when is a function of a base class

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java (hunk 3 above)
Comment (spilledyear) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java]: ok

File: spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java (hunk 2 above)
Comment (spilledyear) [reply spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java]: ok

File: spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java (hunk 1 above)
Comment (spilledyear) [reply spring/src/main/java/io/seata/spring/annotation/datasource/DataSourceProxyHolder.java]: ok

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java (hunk 3 above)
Comment (spilledyear) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java]: ok

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java [hunk 5]
```diff
@@ -53,15 +53,14 @@ public SeataAutoDataSourceProxyAdvice(String dataSourceProxyMode) {
...
-                && dataSourceProxyMode != RootContext.getBranchType()) {
+        if (!RootContext.requireGlobalLock() && dataSourceProxyMode != RootContext.getBranchType()) {
             return invocation.proceed();
         }
 
         Method method = invocation.getMethod();
         Object[] args = invocation.getArguments();
         Method m = BeanUtils.findDeclaredMethod(dataSourceProxyClazz, method.getName(), method.getParameterTypes());
-        if (m != null) {
+        if (m != null && method.getDeclaringClass().isAssignableFrom(DataSource.class)) {
```
Comment (a364176773) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java:63]: DruidDataSource isAssignableFrom DataSource Condition is not established

Comment (spilledyear) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyAdvice.java]: fixed

**Review by a364176773 — APPROVED** [review]
LGTM

**Review by xingfudeshi — APPROVED** [review]
LGTM.

**Pull request merged**